File: app/src/main/java/com/convergencelabstfx/smartdrone/viewmodels/DroneViewModel.java
package com.convergencelabstfx.smartdrone.viewmodels;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.convergencelabstfx.keyfinder.MusicTheory;
import com.convergencelabstfx.keyfinder.ParentScale;
import com.convergencelabstfx.keyfinder.Scale;
import com.convergencelabstfx.keyfinder.harmony.VoicingTemplate;
import com.convergencelabstfx.keyfinder.keypredictor.KeyPredictor;
import com.convergencelabstfx.keyfinder.keypredictor.KeyPredictorListener;
import com.convergencelabstfx.keyfinder.keypredictor.Phrase;
import com.convergencelabstfx.keyfinder.keypredictor.PhrasePredictor;
import com.convergencelabstfx.smartdrone.models.ChordConstructor;
import com.convergencelabstfx.smartdrone.models.MidiPlayer;
import com.convergencelabstfx.smartdrone.models.NoteProcessor;
import com.convergencelabstfx.smartdrone.models.NoteProcessorObserver;
import com.convergencelabstfx.smartdrone.models.ScaleConstructor;
import com.convergencelabstfx.smartdrone.models.SignalProcessorKt;
import com.convergencelabstfx.smartdrone.models.SignalProcessorObserver;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import timber.log.Timber;


/**
 * Drone pipeline:
 * 1) Process signal; TarsosDSP signal processing
 * 2) Process note data; analyze note probability, how long note was heard, etc... (filter out noise)
 * 3) Determine key; analyze processed notes (KeyPredictor)
 * 4) Determine chord;
 * 5) Play chord; MidiDriver
 * 6) Notify UI;
 */

public class DroneViewModel extends ViewModel {

    // todo: remove; just a place holder field

//    private Scale mCurScale;

    private SignalProcessorKt mSignalProcessor = new SignalProcessorKt();

    private NoteProcessor mNoteProcessor = new NoteProcessor();

    private KeyPredictor mKeyPredictor;

    private List<ParentScale> mParentScales = new ArrayList<>();
    private List<VoicingTemplate> mVoicingTemplates = new ArrayList<>();

    private ChordConstructor mChordConstructor = new ChordConstructor();

    private MidiPlayer mMidiPlayer = new MidiPlayer();

    public MutableLiveData<Boolean> mDroneIsActive = new MutableLiveData<>();

    public MutableLiveData<Integer> mDetectedNote = new MutableLiveData<>();
    public MutableLiveData<Integer> mUndetectedNote = new MutableLiveData<>();

    public MutableLiveData<Integer> mDetectedKey = new MutableLiveData<>();

    public MutableLiveData<Scale> mCurScale = new MutableLiveData<>();

    public DroneViewModel() {
        testMethod_setupKeyPredictor();
        testMethod_setupChordConstructor();
        testMethod_setupMidiPlayer();
        testMethod_setupParentScales();
        testMethod_setupVoicingTemplates();
        initPipeline();
    }

    public void startDrone() {
        mSignalProcessor.start();
        mMidiPlayer.start();
        mDetectedKey.setValue(-1);
        mDroneIsActive.setValue(true);
    }

    public void stopDrone() {
        mSignalProcessor.stop();
        mMidiPlayer.stop();
        mDetectedNote.setValue(-1);
        mDetectedKey.setValue(-1);
        mDroneIsActive.setValue(false);
    }

    public void setKeyChange(int key) {
        if (isRunning() && mDetectedKey.getValue() != null && key != mDetectedKey.getValue()) {
            mChordConstructor.setKey(key);
            mDetectedKey.setValue(key);
            mMidiPlayer.clear();
            mMidiPlayer.playChord(mChordConstructor.makeVoicing());
        }
    }

    public boolean isRunning() {
        return mDroneIsActive.getValue() != null && mDroneIsActive.getValue();
    }

    // todo: figure how this should work
    public String getCurModeName() {
        return null;
    }

    public List<ParentScale> getParentScales() {
        return mParentScales;
    }

    public List<VoicingTemplate> getVoicingTemplates() {
        return mVoicingTemplates;
    }

    public void setScale(Scale scale) {
        mCurScale.setValue(scale);
        // todo: update playback
        mChordConstructor.setMode(scale.getIntervals());
        if (mMidiPlayer.hasActiveNotes()) {
            mMidiPlayer.clear();
            mMidiPlayer.playChord(mChordConstructor.makeVoicing());
        }
    }

    public void setVoicingTemplate(VoicingTemplate template) {
        mChordConstructor.setTemplate(template);
        if (mMidiPlayer.hasActiveNotes()) {
            mMidiPlayer.clear();
            mMidiPlayer.playChord(mChordConstructor.makeVoicing());
        }
    }

    // todo: just a method for development purposes; should delete later
    private void testMethod_setupKeyPredictor() {
        // todo: find a better place for this eventually
        final Phrase mOctavePhrase = new Phrase();
        mOctavePhrase.addNote(0);
        mOctavePhrase.addNote(12);
        final PhrasePredictor predictor = new PhrasePredictor();
        predictor.targetPhrase = mOctavePhrase;

        mKeyPredictor = predictor;
    }

    private void testMethod_setupVoicingTemplates() {
        final VoicingTemplate template = new VoicingTemplate();
        template.addBassTone(0);
        template.addChordTone(0);
        template.addChordTone(4);
        template.addChordTone(9);
        mVoicingTemplates.add(template);

        final VoicingTemplate template2 = new VoicingTemplate();
        template2.addBassTone(0);
        template2.addChordTone(6);
        template2.addChordTone(9);
        template2.addChordTone(11);
        mVoicingTemplates.add(template2);

        final VoicingTemplate template3 = new VoicingTemplate();
        template3.addBassTone(0);
        template3.addChordTone(3);
        template3.addChordTone(6);
        template3.addChordTone(9);
        mVoicingTemplates.add(template3);
    }

    private void testMethod_setupChordConstructor() {
        final List<Integer> mode = new ArrayList<>();
        mode.add(0);
        mode.add(2);
        mode.add(3);
        mode.add(5);
        mode.add(7);
        mode.add(9);
        mode.add(10);

        final VoicingTemplate template = new VoicingTemplate();
        template.addBassTone(0);
        template.addBassTone(4);

        template.addChordTone(1);
        template.addChordTone(2);
        template.addChordTone(4);
        template.addChordTone(8);

        mChordConstructor.setMode(mode);
        mChordConstructor.setKey(0);
        mChordConstructor.setTemplate(template);
        mChordConstructor.setBounds(36, 60, 48, 72);
    }

    private void testMethod_setupMidiPlayer() {
        // todo: yeah it's hardcoded for now
        mMidiPlayer.setPlugin(48);
    }

    private void testMethod_setupParentScales() {
        final ParentScale majorScale = ScaleConstructor.makeParentScale(
                "Major Scale",
                Arrays.asList(MusicTheory.MAJOR_SCALE_SEQUENCE),
                Arrays.asList(MusicTheory.MAJOR_MODE_NAMES)
        );
        final ParentScale melodicMinor = ScaleConstructor.makeParentScale(
                "Melodic Minor",
                Arrays.asList(MusicTheory.MELODIC_MINOR_SCALE_SEQUENCE),
                Arrays.asList(MusicTheory.MELODIC_MINOR_MODE_NAMES)
        );
        mParentScales.add(majorScale);
        mParentScales.add(melodicMinor);

        mCurScale.setValue(mParentScales.get(0).getScaleAt(0));
    }

    // todo: gotta figure out exactly where a note index should turn into a note object
    // todo: make consist naming (observer/listener, notify/handle, onKeyPrediction etc...)
    private void initPipeline() {
        mSignalProcessor.addPitchListener(new SignalProcessorObserver() {
            @Override
            public void handlePitchResult(int pitch, float probability, boolean isPitched) {
                mDetectedNote.setValue(pitch);
                mNoteProcessor.onPitchDetected(pitch, probability, isPitched);
            }
        });

        mNoteProcessor.addNoteProcessorListener(new NoteProcessorObserver() {
            @Override
            public void notifyNoteDetected(int note) {
                Timber.i("Detected: " + note);
                mKeyPredictor.noteDetected(note);

            }

            @Override
            public void notifyNoteUndetected(int note) {
                mKeyPredictor.noteUndetected(note);
//                mUndetectedNote.setValue(note);
            }
        });

        mKeyPredictor.addListener(new KeyPredictorListener() {
            @Override
            public void notifyKeyPrediction(int newKey) {
                Timber.i("key: %s", newKey);
                mChordConstructor.setKey(newKey);
                mDetectedKey.setValue(newKey);
                // todo: implement
                mMidiPlayer.clear();
                mMidiPlayer.playChord(mChordConstructor.makeVoicing());
                Timber.i("newKey: %s", newKey);
            }
        });

    }

}

File: app/src/main/java/com/convergencelabstfx/smartdrone/viewmodels/DroneViewModel.kt
package com.convergencelabstfx.smartdrone.viewmodels

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import cn.sherlock.com.sun.media.sound.SF2Soundbank
import com.convergencelabstfx.keyfinder.ParentScale
import com.convergencelabstfx.keyfinder.Scale
import com.convergencelabstfx.keyfinder.harmony.VoicingTemplate
import com.convergencelabstfx.keyfinder.keypredictor.KeyPredictor
import com.convergencelabstfx.smartdrone.R
import com.convergencelabstfx.smartdrone.database.DroneDatabase.Companion.getDatabase
import com.convergencelabstfx.smartdrone.database.DroneRepository
import com.convergencelabstfx.smartdrone.database.VoicingTemplateEntity
import com.convergencelabstfx.smartdrone.models.*
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

/**
 * Drone pipeline:
 * 1) Process signal; TarsosDSP signal processing
 * 2) Process note data; analyze note probability, how long note was heard, etc... (filter out noise)
 * 3) Determine key; analyze processed notes (KeyPredictor)
 * 4) Determine chord;
 * 5) Play chord; MidiDriver
 * 6) Notify UI;
 */
// todo: clean up the java-isms; and clean up the MutableLiveData
class DroneViewModel(application: Application) : AndroidViewModel(application) {

    private val repository: DroneRepository

    var curTemplate: MutableLiveData<VoicingTemplate> = MutableLiveData()
    var curScale = MutableLiveData<Scale>()

    private val signalProcessor = SignalProcessorKt()
    private val noteProcessor = NoteProcessor()
    private val chordConstructor = VoicingConstructor()
    private var keyPredictor: KeyPredictor


    private var midiPlayer: MidiPlayer
    private val mParentScales: List<ParentScale>


    private val mVoicingTemplates: MutableList<VoicingTemplate> = ArrayList()
    @JvmField
    var mDroneIsActive = MutableLiveData<Boolean?>()
    @JvmField
    var mDetectedNote = MutableLiveData<Int>()
    var mUndetectedNote = MutableLiveData<Int>()
    @JvmField
    var mDetectedKey = MutableLiveData<Int?>()

    init {
        val templateDao = getDatabase(application, viewModelScope).voicingTemplateDao()
        val resources = application.resources
        val sharedPreferences = application.getSharedPreferences(resources.getString(R.string.file_key), Context.MODE_PRIVATE)
        repository = DroneRepository(templateDao, sharedPreferences, resources)

        val sfMidiPlayer = MidiPlayer()
        sfMidiPlayer.setPlugin(repository.getMidiPlugin())

        val sf = SF2Soundbank(application.assets.open(repository.getSf2FileName()))
        sfMidiPlayer.sf2 = sf
        sfMidiPlayer.setPlugin(repository.getMidiPlugin())
        midiPlayer = sfMidiPlayer

        mParentScales = repository.getParentScales()
        setScale(mParentScales[repository.getParentScaleIx()].getScaleAt(repository.getModeIx()))
        chordConstructor.bounds = repository.getVoicingBounds()

        keyPredictor = repository.getKeyPredictor()
        setVoicingTemplate(repository.getCurTemplate())

        initPipeline()

    }

    fun startDrone() {
        signalProcessor.start()
        midiPlayer.start()
        mDetectedKey.value = -1
        mDroneIsActive.value = true
    }

    fun stopDrone() {
        signalProcessor.stop()
        midiPlayer.stop()
        mDetectedNote.value = -1
        mDetectedKey.value = -1
        mDroneIsActive.value = false
    }

    fun setKeyChange(key: Int) {
        if (isRunning && mDetectedKey.value != null && key != mDetectedKey.value) {
            chordConstructor.key = key
            mDetectedKey.value = key
            midiPlayer.clear()
            midiPlayer.playChord(chordConstructor.makeVoicing())
        }
    }

    fun toggleBassTone(degree: Int) {
        val template = curTemplate.value
        val note: Int
        if (template?.bassTones!!.contains(degree)) {
            note = chordConstructor.removeBassTone(degree)
            if (midiPlayer.isRunning) {
                midiPlayer.stopNote(note)
            }
        }
        else {
            note = chordConstructor.addBassTone(degree)
            if (midiPlayer.isRunning) {
                midiPlayer.playNote(note)
            }
        }
        setVoicingTemplate(template)
    }

    fun toggleChordTone(degree: Int) {
        val template = curTemplate.value
        val note: Int
        if (template?.chordTones!!.contains(degree)) {
            note = chordConstructor.removeChordTone(degree)
            if (midiPlayer.isRunning) {
                midiPlayer.stopNote(note)
            }
        }
        else {
            note = chordConstructor.addChordTone(degree)
            if (midiPlayer.isRunning) {
                midiPlayer.playNote(note)
            }
        }
        setVoicingTemplate(template)
    }

    fun saveScaleIxs(parentIx: Int, modeIx: Int) {
        setScale(mParentScales[parentIx].getScaleAt(modeIx))
        repository.saveScaleIxs(parentIx, modeIx)
    }

    val isRunning: Boolean
        get() = mDroneIsActive.value != null && mDroneIsActive.value!!

    val parentScales: List<ParentScale>
        get() = mParentScales

    val voicingTemplates: List<VoicingTemplate>
        get() = mVoicingTemplates

    fun setScale(scale: Scale) {
        curScale.value = scale
        // todo: update playback
        chordConstructor.mode = scale.intervals
        if (midiPlayer.hasActiveNotes()) {
            midiPlayer.clear()
            midiPlayer.playChord(chordConstructor.makeVoicing())
        }
    }

    fun setVoicingTemplate(template: VoicingTemplate) {
        chordConstructor.template = template
        curTemplate.value = template
        repository.saveCurTemplate(template)
//        if (isRunning) {
//            chordConstructor.makeVoicing()
////            if (restartPlayback) {
//                midiPlayer.clear()
//                midiPlayer.playChord(chordConstructor.curVoicing)
////            }
//        }
    }

    fun insertVoicingTemplate(template: VoicingTemplate?) = viewModelScope.launch {
        val templateEntity = VoicingTemplateEntity(template!!)
        repository.insert(templateEntity)
    }

    // todo: gotta figure out exactly where a note index should turn into a note object
    // todo: make consist naming (observer/listener, notify/handle, onKeyPrediction etc...)
    private fun initPipeline() {
        signalProcessor.addPitchListener(SignalProcessorObserver { pitch, probability, isPitched ->
            mDetectedNote.value = pitch
            noteProcessor.onPitchDetected(pitch, probability, isPitched)
        })
        noteProcessor.addNoteProcessorListener(object : NoteProcessorObserver {
            override fun notifyNoteDetected(note: Int) {
                Timber.i("Detected: $note")
                keyPredictor.noteDetected(note)
            }

            override fun notifyNoteUndetected(note: Int) {
                keyPredictor.noteUndetected(note)
                //                mUndetectedNote.setValue(note);
            }
        })
        keyPredictor.addListener { newKey ->
            Timber.i("key: %s", newKey)
            chordConstructor.key = newKey
            mDetectedKey.value = newKey
            // todo: implement
            midiPlayer.clear()
            midiPlayer.playChord(chordConstructor.makeVoicing())
            Timber.i("newKey: %s", newKey)
        }
    }

}